Add compute resources role

ephracis opened the pull request:
Create a role for managing compute resources.

ephracis committed:
Add compute resources role

Create a role for managing compute resources.

ephracis commented:
I created this for a client and figured it would be nice to have it upstream.

ephracis committed:
Switch from Satellite to Foreman

Rename some stuff to use the upstream name "foreman" instead of
"satellite".

mdellweg commented on roles/compute_resources/README.md:
```diff
@@ -0,0 +1,44 @@
+- `description`: Description of the compute resource
+- `provider`: Compute resource provider. Required if *state=present_with_defaults*.
+- `provider_params`: Parameter specific to compute resource provider. Required if *state=present_with_defaults*.
+- `updated_name`: New compute resource name. Setting this will make the role not idempotent.
```
Not sure this is really helpful in a role. Maybe renaming is more for ad-hoc ansible calls.

mdellweg commented on roles/compute_resources/README.md:
```diff
@@ -0,0 +1,44 @@
+Example Playbooks
+-----------------
+
+Create a lifecycle environment path with three environments: Library -> Dev -> Test -> Prod
```
Not as such...

mdellweg commented on roles/compute_resources/README.md:
```diff
@@ -0,0 +1,44 @@
+              url: "vcenter.example.com"
+              user: "administrator@vsphere.local"
+              password: "changeme"
+              datacenter: "ha-datacenter"
```
I know, we talked about (operating-system/compute-resource -) images. Would it be a good architecture to define the images as a list in this Compute-resource dict?

ephracis replied:
I definitely see some logical reasons for this since images are tightly coupled with compute resources (must have the latter to do the former). Should I add it?

mdellweg replied:
I'm not in the position to decide, but i'd rather say yes.
cc @ehelms

ehelms replied:
I agree with adding it.

ephracis replied:
Done.

ephracis committed:
Remove updated_name

This parameter makes no sense in a role since it is more useful in a
one-off ad-hoc run.

ephracis committed:
Fix copy-pasted example text

The example playbook in the Readme contained some copy-pasted text,
changing it to actually describe the playbook.

mdellweg commented on roles/compute_resources/README.md:
```diff
@@ -21,7 +21,7 @@ The following fields are optional and will be omitted by default:
 -----------------
 
-Create a lifecycle environment path with three environments: Library -> Dev -> Test -> Prod
+Create a compute resource for vSphere.
```
It handles a list of cr right?

ephracis replied:
Yes, but the example playbook creates just a single one. Should I add more items to the list in the example?

mdellweg replied:
Right. Sorry. In that case, no.

ephracis committed:
Add task to create images in CR role

Add a second task to the `compute_resources` role which creates
associated images for that CR, and include an example of this in the
role Readme.

ephracis commented:
Should there be a changelog fragment for this?

mdellweg commented:
I only know about automatic changelogs for new modules.
Can you add tests please?

ephracis commented:
> I only know about automatic changelogs for new modules. Can you add tests please?

Sure thing. But it seems I must switch this over to another machine since I cannot get `make test-setup` to work on my Mac.

ephracis committed:
Add tests for CR role

mdellweg commented on tests/fixtures/apidoc/compute_resources_role.json:
```diff
@@ -0,0 +1 @@
+katello.json
```
Why is this read? I hope it's a symlink. And think i remember github was way better than that to show it.

ephracis replied:
Yes, that is a symlink.

mdellweg commented on tests/test_playbooks/compute_resources_role.yml:
```diff
@@ -0,0 +1,57 @@
+    - role: compute_resources
+      vars:
+        foreman_organization: "Test Organization"
+        foreman_compute_resources:
```
I think, we want to see images here. :D

ephracis replied:
Oh, I forgot. Of course!

ephracis committed:
Add images to CR tests

Create an image when creating CR via role. Also remove compute_resource
param from Readme since this is infered via the attachment now.

mdellweg commented on tests/test_playbooks/compute_resources_role.yml:
```diff
@@ -37,6 +44,14 @@
+                user_data: true
+                image_username: "root"
+                image_password: "changeme"
+                uuid: "rhel-8.4-template"
```
:+1:

mdellweg commented on tests/test_playbooks/fixtures/compute_resources_role-0.yml:
```diff
@@ -70,9 +70,11 @@ interactions:
   response:
     body:
-      string: "{\n  \"total\": 1,\n  \"subtotal\": 0,\n  \"page\": 1,\n  \"per_page\":
+      string: "{\n  \"total\": 2,\n  \"subtotal\": 1,\n  \"page\": 1,\n  \"per_page\":
```
This recording looks like the "things" already existed...

ephracis replied:
Yes, I did a new record against the existing Foreman after adding images. Might be that?

Anyway, I can remove the CR and do a re-record. If that's what you mean with "a clean start"?

ephracis replied:
I removed the files and reran `make record_compute_resources_role`.

I had some issues getting the centos7-provision-nightly machine up with forklift today, so I had to change to katello 4.1. That's why some version numbers differ.

mdellweg reviewed (commented):
I think, we don't really test idempotence for roles, but can you please rerecord with a clean start?
Also this deserves a changelog fragment.
Everything else looks great to me.

ephracis committed:
Add changelog fragment

ephracis committed:
Re-record tests for CR role
